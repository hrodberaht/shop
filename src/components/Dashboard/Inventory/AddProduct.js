import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { reset } from 'redux-form';
import ConnectedAddProductForm from './AddProductForm';
import { addProduct } from '../../../store/products/actionCreators';
import { getAuthToken } from '../../../store/auth/selectors';

export class AddProduct extends Component {
  submit = (values) => {
    const { addProductToDB, token, clearForm } = this.props;
    addProductToDB(values, token);
    clearForm('addProduct');
  };

  render() {
    return (
      <div className="add-product-inventory">
        <ConnectedAddProductForm onSubmit={this.submit} />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  token: getAuthToken(state),
});

export default connect(
  mapStateToProps,
  {
    addProductToDB: addProduct,
    clearForm: reset,
  },
)(AddProduct);

AddProduct.propTypes = {
  token: PropTypes.string.isRequired,
  addProductToDB: PropTypes.func.isRequired,
  clearForm: PropTypes.func.isRequired,
};
