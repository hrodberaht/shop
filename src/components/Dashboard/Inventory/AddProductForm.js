import React, { Component } from 'react';
import { Field, reduxForm } from 'redux-form';
import PropTypes from 'prop-types';
import * as yup from 'yup';
import { connect } from 'react-redux';

const schema = yup.object().shape({
  name: yup.string().required(),
  type: yup.string().required(),
  price: yup.number().required(),
  inStock: yup.number().required(),
});

const asyncValidate = values => schema.validate(values, { abortEarly: false }).catch((err) => {
  const errorsForm = {};
  err.inner.forEach((error) => {
    errorsForm[error.path] = error.type;
  });
  throw errorsForm;
});

export class AddProductForm extends Component {
  renderField = ({ input, type, meta: { touched, error } }) => (
    <div>
      <div>
        <input className="form-input" {...input} type={type} />
        {touched && error && <p className="error-text">{error}</p>}
      </div>
    </div>
  );

  render() {
    const { handleSubmit, pristine } = this.props;

    return (
      <div className="product-form">
        <form onSubmit={handleSubmit}>
          <label htmlFor="name">
            <span>Name:</span>
            <Field id="name" name="name" component={this.renderField} type="text" />
          </label>
          <label htmlFor="type">
            <span>Type:</span>
            <Field id="type" name="type" component={this.renderField} type="text" />
          </label>
          <label htmlFor="price">
            <span>Price:</span>
            <Field id="price" name="price" component={this.renderField} type="number" />
          </label>
          <label htmlFor="inStock">
            <span>Stock:</span>
            <Field id="inStock" name="inStock" component={this.renderField} type="number" />
          </label>
          <button className="btn btn-submit" type="submit" disabled={pristine}>
            Add
          </button>
        </form>
      </div>
    );
  }
}

const mapStateToProps = (state, ownProps) => ({
  initialValues: ownProps.product,
  form: ownProps.product.id,
});

export default connect(mapStateToProps)(
  reduxForm({
    asyncValidate,
    enableReinitialize: true,
    form: 'addProduct',
  })(AddProductForm),
);

AddProductForm.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
};
