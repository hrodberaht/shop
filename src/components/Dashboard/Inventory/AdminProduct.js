import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import ConnectedAddProductForm from './AddProductForm';
import { removeProduct, updateProduct } from '../../../store/products/actionCreators';
import { getAuthToken } from '../../../store/auth/selectors';

export class AdminProduct extends Component {
  state = {
    toggleEdit: false,
  };

  handleEditClick = () => {
    this.setState(state => ({
      toggleEdit: !state.toggleEdit,
    }));
  };

  handleRemoveClick = (id) => {
    const { removeProd, token } = this.props;
    removeProd(id, token);
  };

  submit = (values) => {
    const { updateProductInDB, token } = this.props;
    updateProductInDB(values, token);
  };

  render() {
    const {
      product,
      product: {
        id, name, type, price, inStock, remove,
      },
    } = this.props;
    return (
      <React.Fragment>
        {this.state.toggleEdit ? (
          <td colSpan="6">
            <ConnectedAddProductForm product={product} onSubmit={this.submit} />
            <button
              id="closeButton"
              className="btn btn-primary"
              type="button"
              onClick={this.handleEditClick}
            >
              Close
            </button>
          </td>
        ) : (
          <React.Fragment>
            <td>{name}</td>
            <td>{type}</td>
            <td>{+price}</td>
            <td>{+inStock}</td>
            <td>
              <button
                id="editButton"
                className="btn btn-primary"
                type="button"
                onClick={this.handleEditClick}
              >
                Edit
              </button>
            </td>
            <td>
              <button
                id="removeButton"
                className="btn btn-danger"
                type="button"
                onClick={() => this.handleRemoveClick(id)}
                disabled={remove}
              >
                X
              </button>
            </td>
          </React.Fragment>
        )}
      </React.Fragment>
    );
  }
}
const mapStateToProps = state => ({
  token: getAuthToken(state),
});

export default connect(
  mapStateToProps,
  {
    removeProd: removeProduct,
    updateProductInDB: updateProduct,
  },
)(AdminProduct);

AdminProduct.propTypes = {
  product: PropTypes.shape({
    name: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
    price: PropTypes.string.isRequired,
    inStock: PropTypes.string.isRequired,
  }).isRequired,
  removeProd: PropTypes.func.isRequired,
  token: PropTypes.string.isRequired,
  updateProductInDB: PropTypes.func.isRequired,
};
