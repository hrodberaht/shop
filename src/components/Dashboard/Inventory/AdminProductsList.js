import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import ConnectedAdminProduct from './AdminProduct';
import { getProductsAll, getProductsLoaded } from '../../../store/products/selectors';
import { fetchProducts } from '../../../store/products/actionCreators';
import { getAuthToken } from '../../../store/authenticate/selectors';

export class AdminProductsList extends Component {
  componentDidMount() {
    const { getProducts, token } = this.props;
    getProducts(token);
  }

  render() {
    const { products, loaded } = this.props;
    if (!loaded) return <p>Loading...</p>;
    return (
      <div className="admin-product__list">
        <table>
          <tbody>
            <tr>
              <th>Name:</th>
              <th>Image:</th>
              <th>Type:</th>
              <th>Price:</th>
              <th>Stock:</th>
              <th>Edit:</th>
              <th>Remove:</th>
            </tr>
            {products.map(product => (
              <tr key={product.id}>
                <ConnectedAdminProduct product={product} />
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    );
  }
}
const mapStateToProps = state => ({
  products: getProductsAll(state),
  token: getAuthToken(state),
  loaded: getProductsLoaded(state),
});

export default connect(
  mapStateToProps,
  {
    getProducts: fetchProducts,
  },
)(AdminProductsList);

AdminProductsList.propTypes = {
  products: PropTypes.arrayOf(PropTypes.object).isRequired,
  getProducts: PropTypes.func.isRequired,
  token: PropTypes.string.isRequired,
  loaded: PropTypes.bool.isRequired,
};
