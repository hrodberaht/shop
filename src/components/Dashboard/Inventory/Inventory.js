import React, { Component } from 'react';
import ConnectedAdminProductsList from './AdminProductsList';
import ConnectedAddProduct from './AddProduct';

export default class Inventory extends Component {
  state = {
    toggleAddProduct: false,
  };

  handleClick = () => {
    this.setState(state => ({
      toggleAddProduct: !state.toggleAddProduct,
      toggleModal: !state.toggleModal,
    }));
  };

  render() {
    const { toggleAddProduct } = this.state;
    return (
      <div className="inventory">
        {toggleAddProduct ? (
          <button className="btn btn-danger" type="button" onClick={this.handleClick}>
            <span>Close</span>
          </button>
        ) : (
          <button className="btn btn-submit" type="button" onClick={this.handleClick}>
            <span>Add new item</span>
          </button>
        )}

        {toggleAddProduct && <ConnectedAddProduct />}
        <hr />
        <ConnectedAdminProductsList />
      </div>
    );
  }
}
