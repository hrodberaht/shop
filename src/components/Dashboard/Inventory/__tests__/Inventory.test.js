import Inventory from '../Inventory';

const setup = buildSetup(Inventory);

describe('<Inventory />', () => {
  it('should render witout crash', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });
  it('should show AddProduct if toggle is true', () => {
    const { wrapper } = setup();
    wrapper.setState({ toggleAddProduct: true });
    expect(wrapper).toMatchSnapshot();
  });
  it('should change state after click button ', () => {
    const { wrapper } = setup();
    wrapper.find('button').simulate('click');

    expect(wrapper.state().toggleAddProduct).toBe(true);
  });
});
