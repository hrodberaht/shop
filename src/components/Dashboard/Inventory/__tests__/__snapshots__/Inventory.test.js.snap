// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Inventory /> should render witout crash 1`] = `
<div
  className="inventory"
>
  <button
    className="btn btn-primary"
    onClick={[Function]}
    type="button"
  >
    <span>
      Add new item
    </span>
  </button>
  <hr />
  <Connect(AdminProductsList) />
</div>
`;

exports[`<Inventory /> should show AddProduct if toggle is true 1`] = `
<div
  className="inventory"
>
  <button
    className="btn btn-primary"
    onClick={[Function]}
    type="button"
  >
    <span>
      Close
    </span>
  </button>
  <Connect(AddProduct) />
  <hr />
  <Connect(AdminProductsList) />
</div>
`;
